### ISSUE TITLE
Disconnected volume doesn't reappear on reconnect

### ISSUE THREAD

**Issue by jandubois** [docker/for-mac#773]
### Expected behavior

I have a volume from the NAS mounted on the Mac as `/Volumes/NAS`. It is bound into a container via `docker run -v /Volumes/NAS:/mnt/nas` and works as expected. When the volume is ejected, e.g. to restart the NAS, accessing `/mnt/nas` throws an error `cannot access /mnt/nas: No such file or directory`. This is also expected.

When the `/Volumes/NAS` on the host is recreated, I would expect the `/mnt/nas` inside the container to start working again.
### Actual behavior

Once the host volume has disappeared, the mount point inside the container remains disconnected even when the host directory reappears.
### Information

Docker for Mac: version: 1.12.2-rc1-beta27 (179c18c)
OS X: version 10.11.6 (build: 15G1004)
logs: /tmp/C386DC53-7999-448D-98A6-5AB810275D79/20161009-122154.tar.gz
[OK]     docker-cli
[OK]     virtualization kern.hv_support
[OK]     menubar
[OK]     moby-syslog
[OK]     dns
[OK]     disk
[OK]     system
[OK]     app
[OK]     osxfs
[OK]     virtualization VT-X
[OK]     db
[OK]     slirp
[OK]     logs
[OK]     env
[OK]     vmnetd
[OK]     moby-console
[OK]     moby
[OK]     driver.amd64-linux
### Steps to reproduce the behavior

Connect `/Volumes/NAS` on the host.

```
$ docker run -it --rm -v /Volumes/NAS:/mnt/nas debian:jessie bash
root@713e7508805e:/# ls /mnt/nas
#snapshot  Documents  Library  Reference  Software  Video
```

Eject NAS on host.

```
root@713e7508805e:/# ls /mnt/nas
ls: cannot access /mnt/nas: No such file or directory
```

Reconnect NAS on host.

```
root@713e7508805e:/# ls /mnt/nas
ls: cannot access /mnt/nas: No such file or directory
```

**Comment by jandubois** [comment]
Sorry, didn't find https://docs.docker.com/docker-for-mac/osxfs/#mounts until after filing this.  I guess the behavior is covered by

> Mounting/unmounting OS X volumes that are also bind mounted into containers may result in unexpected behavior in those containers.

**Comment by dsheets** [comment]
Thanks for the report! I'm reopening this issue to track the generally poor user experience you mention. There are a couple of implementation options to make this better:

1. Pop up a message saying something like "container X has host volume Y mounted but host volume Y is being unmounted. [Force unmount] [Stop container X and retry] [Abort eject]"
2. Export the OS X mount table into the VM and update it to reflect changes in OS X.

It's also possible to have the container mount only treated as a reference to the OS X mount and ensure that, if the OS X mount disappears, the container mount's cache is properly cleared. I am very hesitant to go down this route, though, as the container mount point would no longer behave as other Linux file systems as inode maps, `statfs` output, etc could change.

What behavior do you think would be most easily understandable/natural? Does your use case require seamless behavior? What if we provided some way to automatically start/stop containers when a specific volume is mounted in OS X?

**Comment by jandubois** [comment]
# TL;DR

I think a mechanism to automatically restart containers when a mounted volume reappears is the most robust option.

# Popup message

Note that a clean disconnect by explicit "Eject" is probably the exception. I just used it as a shortcut during testing. My real use cases are:

1. NAS does software auto-update and reboots. Potentially in some configurations also NAS goes to sleep and disconnects.

2. Laptop moved to a different location without route to NAS. This may happen while laptop is sleeping.

Both cases rely on some automount mechanism to reconnect.

Either way, you don't have a chance to display a popup during disconnect. Also this may happen on a headless Mac mini, so there will be nobody around to click buttons.

# My current workaround

I couldn't get automount to work on OS X without storing username/password in cleartext files, which I find unacceptable. Instead I wrote a little AppleScript that runs every minute via `launchd` that tries to remount volumes that have gone missing.

```
use scripting additions

try
	do shell script "ping -c 1 nas | grep -q icmp"
	mount("Documents")
	mount("Library")
	mount("Media")
end try

on mount(volumeName)
	tell application "Finder"
		if not (disk volumeName exists) then
			mount volume "smb://nas._smb._tcp.local/" & volumeName
			if disk volumeName exists then
				do shell script "/usr/local/bin/docker restart $(/usr/local/bin/docker ps -q --filter volume=/Volumes/" & volumeName & ")"
			end if
		end if
	end tell
end mount
```

This takes advantage of the fact that the Finder can store credentials in the Keychain, and use those automatically during mounting.

I've added a `docker restart` whenever a volume is reconnected. In general a container may continue to do useful work in a degraded state because a mount may be optional (e.g. just one of multiple media sources), so always stopping when a volume goes away is not desirable.

I thought some more about my original expectation that mounts should transparently reconnect, but see now that the software inside the container may not deal with it correctly anyways, so the restart seems like the more robust option.

Given that I currently need my custom automounter solution anyways, adding the `docker restart` in there is good enough for me personally. However, if automounter ever makes uses of encrypted credentials, then having docker doing the restart automatically would be nice.
